README
2015-09-22 David L. Parsley <parsley@linuxjedi.org> version 0.4.10
License: GPL, see COPYING for full text.

NOTE: I no longer run my own dhcp server, so can only give minimal time
to this project. I'll gladly read and accept reasonable patches with
explanations.

omcmd is a scriptable CLI program for performing actions on a running ISC dhcp
server using OMAPI.  It can currently be used to create, remove, modify and 
look up hosts and leases.

Note that these objects are 'stored' in the dhcpd.leases file.

Building: see README-Compiling

Usage: omcmd [-k keyname secret | -K keyfile ] [-s server] [-v] \
  <command> <objtype> <args>

The keyname & secret are the same as the values in the dhcpd.conf file;
the secret is base64 encoded.  If using a 'keyfile', line1=keyname, 
line2=secret.

-v for verbose output (see below)

command is one of:
    lookup - find an object and print  properties
    create - create a new object
    remove - remove an object
    modify - modify an object

objtype is one of:
    host
    lease

For all commands, the first arg should be of the form property=value, and
should be a 'lookup' value (see OMAPI in man dhcpd), i.e.: 
ip-address=10.20.30.40.  For remove, this is the only required argument.

For create & modify, the remaining arguments are more property=value strings
supported properties so far:
    ip-address - e.g. "ip-address=10.20.30.40"
    hardware-type - e.g. "hardware-type=1" (you almost always use this when 
	specifying the hardware-address)
    hardware-address - e.g. "hardware-address=00:10:20:30:40:50"
    state - e.g. "state=free" (you can use this to free an active lease**)
    name - e.g. "name=mylaptop"
    client-hostname (similar)
For the appropriate properties for an object type, see 'man dhcp'.

** dhcp-3.0.1 won't allow leases to be freed dynamically as-is.  I've included
a (short) dhcp-omapifree.patch, that, when applied, will actually free leases
if state is set to 'free'.

For 'lookup', you can leave off remaining args, and it will print what it
finds (in verbose mode).  You can also specify property names you want found,
and it will spit them out one per line (non-verbose mode), or as
property=value strings (verbose). Lookup properties supported are: ends, tstp,
tsfp, cltt, ip-address, client-hostname, name, dhcp-client-identifier,
hardware-address, hardware-type, and state.

Currenlty, only host & lease objects are supported, but adding support for
other known object types shouldn't be hard.

The return value is 255 for invalid usage, otherwise a value of type
isc_result_t.
(see include/isc-dhcp/result.h)

Examples:

Lookup a lease and print everything:

$ ./omcmd -K ~/keyfile -s myserver lookup lease ip-address=10.20.93.155
state=active
ip-address=10.20.93.155
dhcp-client-identifier=01:00:e0:b8:72:8f:90
client-hostname=AU1201
hardware-address=00:e0:b8:72:8f:90
hardware-type=1
ends=Sat Jul 31 15:54:28 2004
tstp=Wed Dec 31 19:00:00 1969
tsfp=Wed Dec 31 19:00:00 1969
cltt=Wed Dec 31 19:00:00 1969

Lookup a lease and just give the hardware-address:

$ ./omcmd -K ~/keyfile -s myserver lookup lease ip-address=10.20.93.155 hardware-address
00:e0:b8:72:8f:90

Change a lease state to free:

$ ./omcmd -K ~/keyfile -s myserver modify lease ip-address=10.20.162.42 state=free

Note: I wasn't able to 'remove' a lease; the best I could figure was to 'free'
it.
Note2: the first prop=value is used to look up the object, all following will
update the object.

Create a new host:

$ ./omcmd -K ~/keyfile -s myserver create host ip-address=10.20.162.242 name=ioreth hardware-type=1 hardware-address=00:e0:b8:50:0f:ab

Note: ip-address is optional.

Delete that host:

$ ./omcmd -K ~/keyfile -s myserver remove host name=ioreth

Feedback & patches to David L. Parsley, parsley@linuxjedi.org


NEWS
omcmd NEWS - history of user-visible changes, 2005-01-03

Version 0.4.8
* Allow statements for hosts (patch from Caciano Machado)
* Support for group objects (patch from Caciano Machado)
* Add -p option to specify port

Version 0.4.7
* Print errors to stdout
* Allow times to be specified (not currently supported in server 3.0.1)


omcmd.c
/* omcmd.c - see COPYING for license
 *  (c)2004 David L. Parsley, parsley@linuxjedi.org
 */

#include <stdarg.h>
#include <stdlib.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <stdio.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <string.h>

#include <omapip/result.h>
#include <dhcpctl/dhcpctl.h>

#include "omcmd.h"

void usage() {
    printf("Usage: omcmd [-k keyname secret | -K keyfile ] [-s server] [-v] \\\n");
    printf("  <command> <objtype> <args>\n");
    exit(-1);
}

void toobig(char *var,char *val) {
    printf("%s too long: %s longer than %d\n",var,val,MAXLEN);
    exit(-1);
}

void exitout(isc_result_t status) {
    if (status) fprintf(stderr, "Error: %s\n", isc_result_totext(status));
    exit(status);
}

int main (int argc, char **argv) {
    /* values for setting */
    dhcpctl_data_string values[16];
    /* value for getting */
    dhcpctl_data_string value=NULL;
    dhcpctl_handle authenticator=dhcpctl_null_handle;
    dhcpctl_handle object=dhcpctl_null_handle;
    dhcpctl_handle connection=dhcpctl_null_handle;
    isc_result_t status;
    const char *algorithm = "hmac-md5";
    char keyname[MAXLEN + 1];
    char secret[MAXLEN + 1];
    char secret_base64[MAXLEN + 1];
    char server[MAXLEN +1];
    char objstring[MAXLEN +1];
    omtype objtype;
    char valstring[MAXLEN +1];
    char **lookupprops;
    int lookups;
    char *defaulthostprops[]={"name","hardware-address","hardware-type",
	"dhcp-client-identifier","ip-address","known","group"};
    int defaulthostnumprops=7;
    char *defaultleaseprops[]={"state","ip-address","dhcp-client-identifier",
	"client-hostname","hardware-address","hardware-type","ends","tstp",
	"tsfp","cltt"};
    int defaultleasenumprops=10;
    char *defaultgroupprops[]={"name"};
    int defaultgroupnumprops=1;
    omprop proptype;
    int seclength=0;
    int argindex=1;
    int valindex=1;
    int i;
    int waitfor=0;
    int port=7911;
    int outputformat=OM_TERSE;
    int slen;
    omcmd command;
    FILE *keyfile;

    object=dhcpctl_null_handle;
    for (valindex=0; valindex<16; valindex++){
	memset(&values[valindex],0,sizeof values[valindex]);
    }
    valindex=0;
    strcpy(server,"127.0.0.1");

    // process options
    while(argindex < argc && argv[argindex][0]=='-'){
	if ((argv[argindex][1] == 0) || ( argv[argindex][2] != 0)) {
	    usage();
	}
	switch(argv[argindex][1]){
	    case 'k':
		if (strlen(argv[++argindex]) < MAXLEN)
		    strcpy(keyname,argv[argindex]);
		else toobig("keyname",argv[argindex]);
		if (strlen(argv[++argindex]) < MAXLEN){
		    seclength=from64tobits(secret,argv[argindex]);
		    if (seclength == -1){
			fprintf(stderr,"Invalid base64 string: %s\n",argv[argindex]);
			exit(-1);
		    }
		}
		else toobig("secret",argv[argindex]);
		argindex++;
		break;
	    case 'K':
		if ((keyfile=fopen(argv[++argindex],"r"))==NULL){
		    fprintf(stderr,"Unable to open key file %s for reading\n",
			argv[argindex]);
		    exit(-1);
		}
		if (fgets(keyname,MAXLEN,keyfile)==NULL){
		    fprintf(stderr,"Unable to read keyname from key file %s\n",
			argv[argindex]);
		    exit(-1);
		}
		slen=strlen(keyname);
		if (keyname[slen-1] != '\n'){
		    toobig("keyname",keyname);
		}
		keyname[slen-1]='\0';
		if (fgets(secret_base64,MAXLEN,keyfile)==NULL){
		    fprintf(stderr,"Unable to read secret from key file %s\n",
			argv[argindex]);
		    exit(-1);
		}
		slen=strlen(secret_base64);
		if (secret_base64[slen-1] != '\n'){
		    toobig("base64 encoded secret",secret_base64);
		}
		secret_base64[slen-1]='\0';
		seclength=from64tobits(secret,secret_base64);
		if (seclength == -1){
		    fprintf(stderr,"Invalid base64 string: %s\n",secret_base64);
		    exit(-1);
		}
		argindex++;
		break;
	    case 'p':
		port = atoi(argv[++argindex]);
		argindex++;
		break;
	    case 's':
		if (strlen(argv[++argindex]) < MAXLEN)
		    strcpy(server,argv[argindex]);
		else toobig("server",argv[argindex]);
		argindex++;
		break;
	    case 'v':
		    outputformat=OM_VERBOSE;
		    argindex++;
		    break;
	    default:
		usage();
	}
    }

    if (argindex>=argc){
	fprintf(stderr, "Missing <command>\n");
	usage();
    }
    if (!(command=stringtocmd(argv[argindex]))){
	fprintf(stderr, "Unknown command: %s\n", argv[argindex]);
	usage();
    }
    if (++argindex>=argc){
	fprintf(stderr, "Missing <objtype>\n");
	usage();
    }
    if (strlen(argv[argindex]) < MAXLEN)
	strcpy(objstring,argv[argindex]);
    else toobig("objtype",argv[argindex]);
    if (++argindex>=argc){
	fprintf(stderr, "Missing required arguments\n");
	usage();
    }
    objtype=stringtoobj(objstring);
    if (!objtype){
	fprintf(stderr, "Unknown or unsupported object type: %s\n",objstring);
	exit(-1);
    }

    /* Start dhcpctl stuff */
    status=dhcpctl_initialize();
    if (status != ISC_R_SUCCESS) {
	fprintf(stderr, "dhcpctl_initialize: %s\n",
	    isc_result_totext(status));
	exitout(status);
    }
    authenticator=dhcpctl_null_handle;
    if (seclength){
        status=dhcpctl_new_authenticator(&authenticator, keyname, algorithm,
	   secret, seclength);
	if (status != ISC_R_SUCCESS) {
	    fprintf(stderr,"dhcpctl_new_authenticator: %s\n",
		isc_result_totext(status));
	    exitout(status);
	}
    }
    connection = dhcpctl_null_handle;
    status = dhcpctl_connect(&connection,server,port,authenticator);
    if (status != ISC_R_SUCCESS) {
	fprintf(stderr,"dhcpctl_connect: %s\n",
	    isc_result_totext(status));
	exitout(status);
    }
    /* Get an object to work with */
    status=dhcpctl_new_object (&object, connection,objstring);
    if (status != ISC_R_SUCCESS) {
	fprintf(stderr,"dhcpctl_new_object: %s\n",
	    isc_result_totext(status));
	exitout(status);
    }
    /* There's always a first attribute=value */
    if (!setvalue(&object,&values[0],argv[argindex])){
	fprintf(stderr,"bad attribute spec: %s\n",argv[argindex]);
	usage();
    }
    switch(command){
	/* For these, we need to open the object first. */
	case oc_lookup:
	case oc_remove:
	case oc_modify:
	    status=dhcpctl_open_object (object, connection, 0);
	    if (status != ISC_R_SUCCESS) {
		fprintf(stderr,"dhcpctl_open_object: %s\n",
		    isc_result_totext(status));
		exitout(status);
	    }
	    dhcpctl_wait_for_completion (object, &status);
	    if (status != ISC_R_SUCCESS) {
		fprintf(stderr,"dhcpctl_wait_for_completion: %s\n",
		    isc_result_totext(status));
		exit (status);
	    }
    }
    if (command == oc_create || command == oc_modify){
	for(argindex++;argindex<argc;argindex++){
	    if (!setvalue(&object,&values[valindex],argv[argindex])){
	        fprintf(stderr,"bad attribute spec: %s\n",argv[argindex]);
		usage();
	    }
	    valindex++;
	}
    }
    switch(command){
	case oc_remove:
	    status=dhcpctl_object_remove(connection, object);
	    waitfor=1;
	    break;
	case oc_create:
	    status=dhcpctl_open_object(object, connection, DHCPCTL_CREATE |
		DHCPCTL_EXCL);
	    waitfor=1;
	    break;
	case oc_modify:
	    status=dhcpctl_object_update(connection, object);
	    waitfor=1;
	    break;
    }
    if (waitfor){
	if (status != ISC_R_SUCCESS) exitout(status);
	dhcpctl_wait_for_completion(object, &status);
	exitout(status);
    }
    /* command == oc_lookup for sure now, all others have exited */
    if (++argindex == argc){
	outputformat=OM_VERBOSE;
	switch(objtype){
	    case obj_host:
		lookupprops=defaulthostprops;
		lookups=defaulthostnumprops;
		break;
	    case obj_lease:
		lookupprops=defaultleaseprops;
		lookups=defaultleasenumprops;
        break;
        case obj_group:
        lookupprops=defaultgroupprops;
        lookups=defaultgroupnumprops;
		break;
	}	
    }
    else{
	lookupprops=&argv[argindex];
	lookups=(argc - argindex);
    }
    for (i=0; i<lookups; i++){
	proptype=stringtoprop(lookupprops[i]);
	if (!proptype){
	    if (outputformat == OM_VERBOSE) printf("%s=",lookupprops[i]);
	    printf("error: unknown or unsupported property type: %s\n",
		lookupprops[i]);
	}
	else{
	    status = dhcpctl_get_value(&value, object, lookupprops[i]);
	    if (status != ISC_R_SUCCESS){
		if (status == ISC_R_NOTFOUND){
		    /* in non-verbose, always one line per specified property */
		    if(outputformat!=OM_VERBOSE) printf("\n");
		}
		else{
		    if (outputformat == OM_VERBOSE) printf("%s=",lookupprops[i]);
		    printf("error: getting %s: %s\n",lookupprops[i],
			isc_result_totext(status));
		}
	    }
	    else {
		if (outputformat == OM_VERBOSE) printf("%s=",lookupprops[i]);
		if(valuetostring(&value, proptype, valstring, MAXLEN +1)==NULL)
		    printf("error: couldn't convert value of %s to a string.\n",
			lookupprops[i]);
		else
		    printf("%s\n",valstring);
	    }
	    dhcpctl_data_string_dereference(&value, MDL);
	}
    }
    exit(ISC_R_SUCCESS);
}


omcmd.h
/* omcmd.h - see COPYING for license.
 * (c)2004 David L. Parsley <parsley@linuxjedi.org>
 */
#include <stdarg.h>
#include <netinet/in.h>

#include <omapip/result.h>
#include <dhcpctl/dhcpctl.h>

/* Max. string length */
#define MAXLEN 255

/* output formats */
#define OM_TERSE 0
#define OM_VERBOSE 1

/* Command constants.
 * zero is error condition */
typedef enum {
    oc_notfound,
    oc_lookup,
    oc_create,
    oc_remove,
    oc_modify
} omcmd;

/* Object properties.
 * Note: all possible object properties aren't implemented. */
typedef enum {
    op_unknown,
    op_state,
    op_ipaddr,
    op_dhcp_client_id,
    op_client_hostname,
    op_hwaddr,
    op_hwtype,
    op_ends,
    op_tstp,
    op_tsfp,
    op_cltt,
    op_name,
    op_statements,
    op_known,
    op_group
} omprop;

/* Possible lease states. */
typedef enum {
    os_unknown,
    os_free,
    os_active,
    os_expired,
    os_released,
    os_abandoned,
    os_reset,
    os_backup,
    os_reserved,
    os_bootp
} omstate;

/* Object types. 
 * Note: not all types implemented. */
typedef enum {
    obj_unknown,
    obj_host,
    obj_lease,
    obj_group
} omtype;

int from64tobits(char*,const char*);
omcmd stringtocmd(char *cmdstring);
omprop stringtoprop(char *propstring);
omtype stringtoobj(char *objstring);
char *statetostring(omstate state);
int setvalue(dhcpctl_handle *object, dhcpctl_data_string *value, char *valspec);
char *valuetostring(dhcpctl_data_string *p_val, omprop proptype, char *buf, int bufsz);


base64.c
/**
 * \file base64.c
 * \brief base-64 conversion routines.
 *
 * \author Eric S. Raymond <esr@snark.thyrsus.com>.
 *
 * For license terms, see the file COPYING in this directory.
 *
 * This base 64 encoding is defined in RFC2045 section 6.8,
 * "Base64 Content-Transfer-Encoding", but lines must not be broken in the
 * scheme used here.
 *
 * Removed unneeded encode function. 2004-07-17 David L. Parsley
 * parsley@linuxjedi.org
 */
#include <ctype.h>
#include "omcmd.h"

#define BAD	-1
static const char base64val[] = {
    BAD,BAD,BAD,BAD, BAD,BAD,BAD,BAD, BAD,BAD,BAD,BAD, BAD,BAD,BAD,BAD,
    BAD,BAD,BAD,BAD, BAD,BAD,BAD,BAD, BAD,BAD,BAD,BAD, BAD,BAD,BAD,BAD,
    BAD,BAD,BAD,BAD, BAD,BAD,BAD,BAD, BAD,BAD,BAD, 62, BAD,BAD,BAD, 63,
     52, 53, 54, 55,  56, 57, 58, 59,  60, 61,BAD,BAD, BAD,BAD,BAD,BAD,
    BAD,  0,  1,  2,   3,  4,  5,  6,   7,  8,  9, 10,  11, 12, 13, 14,
     15, 16, 17, 18,  19, 20, 21, 22,  23, 24, 25,BAD, BAD,BAD,BAD,BAD,
    BAD, 26, 27, 28,  29, 30, 31, 32,  33, 34, 35, 36,  37, 38, 39, 40,
     41, 42, 43, 44,  45, 46, 47, 48,  49, 50, 51,BAD, BAD,BAD,BAD,BAD
};
#define DECODE64(c)  (isascii(c) ? base64val[c] : BAD)

int from64tobits(char *out, const char *in)
/* base 64 to raw bytes in quasi-big-endian order, returning count of bytes */
{
    int len = 0;
    register unsigned char digit1, digit2, digit3, digit4;

    if (in[0] == '+' && in[1] == ' ')
	in += 2;
    if (*in == '\r')
	return(0);

    do {
	digit1 = in[0];
	if (DECODE64(digit1) == BAD)
	    return(-1);
	digit2 = in[1];
	if (DECODE64(digit2) == BAD)
	    return(-1);
	digit3 = in[2];
	if (digit3 != '=' && DECODE64(digit3) == BAD)
	    return(-1); 
	digit4 = in[3];
	if (digit4 != '=' && DECODE64(digit4) == BAD)
	    return(-1);
	in += 4;
	*out++ = (DECODE64(digit1) << 2) | (DECODE64(digit2) >> 4);
	++len;
	if (digit3 != '=')
	{
	    *out++ = ((DECODE64(digit2) << 4) & 0xf0) | (DECODE64(digit3) >> 2);
	    ++len;
	    if (digit4 != '=')
	    {
		*out++ = ((DECODE64(digit3) << 6) & 0xc0) | DECODE64(digit4);
		++len;
	    }
	}
    } while 
	(*in && *in != '\r' && digit4 != '=');

    return (len);
}

/* base64.c ends here */


omfuncs.c
/* omfuncs - functions that would just clutter up omcmd.c ;-)
 * see COPYING for license.
 * (c)2004 David L. Parsley <parsley@linuxjedi.org>
 */

#include <time.h>
#include <string.h>
#include <stdarg.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <stdlib.h>
#include <stdio.h>

#include <omapip/result.h>
#include <dhcpctl/dhcpctl.h>

#include "omcmd.h"

const char *commandstrings[]={
    "lookup",
    "find",
    "create",
    "add",
    "new",
    "remove",
    "delete",
    "modify",
    "update",
    "change",
    "list",
    ""
};
const int commandnum[]={
	1,
	1,
	2,
	2,
	2,
	3,
	3,
	4,
	4,
	4,
	5
};
const char *propstrings[]={
    "state",
    "ip-address",
    "dhcp-client-identifier",
    "client-hostname",
    "hardware-address",
    "hardware-type",
    "ends",
    "tstp",
    "tsfp",
    "cltt",
    "name",
    "statements",
    "known",
    "group"
};
#define OM_NUMPROPS 14
const char *objectstrings[]={
    "host",
    "lease",
    "group"
};
#define OM_NUMOBJTYPES 3
const char *statestrings[]={
    "free",
    "active",
    "expired",
    "released",
    "abandoned",
    "reset",
    "backup",
    "reserved",
    "bootp"
};
#define OM_NUMSTATES 9

omcmd stringtocmd(char *cmdstring){
    int i;
//    for (i=0; i<OM_NUMCOMMANDS; i++)
    for (i=0; commandstrings[i][0] != '\0'; i++)
	if (!strcmp(commandstrings[i],cmdstring)) return((omcmd)commandnum[i]);
    return(oc_notfound); // error
}

omprop stringtoprop(char *propstring){
    int i;
    for (i=0; i<OM_NUMPROPS; i++)
	if (!strcmp(propstrings[i],propstring)) return((omprop)(i+1));
    return(op_unknown);
}

omtype stringtoobj(char *objstring){
    int i;
    for (i=0; i<OM_NUMOBJTYPES; i++)
	if (!strcmp(objectstrings[i],objstring)) return((omtype)(i+1));
    return(obj_unknown);
}

omstate stringtostate(char *statestring){
    int i;
    for (i=0; i<OM_NUMSTATES; i++)
	if (!strcmp(statestrings[i],statestring)) return((omstate)(i+1));
    return(os_unknown);
}

char *statetostring(omstate state){
    if (state>0 && state<OM_NUMSTATES + 1) return((char *)statestrings[state-1]);
    return "unknown";
}

/* valspec is a string of form "var=value" 
 * Return values:
     0 - parse error
    -1 - error setting value
     1 - OK
*/
int setvalue(dhcpctl_handle *p_obj, dhcpctl_data_string *p_val, char *valspec){
    char *var, *val;
    isc_result_t status;
    dhcpctl_handle object;
    dhcpctl_data_string value;
    omprop setproperty;
    omstate setstate;
    struct in_addr convaddr;
    char hwaddr[6];
    int i;
    uint32_t hostint,netint;
    struct tm timestruct;
    time_t timeval;

    object=*p_obj;
    value=*p_val;
    var=valspec;
    val=index(valspec,'=');
    if (val == NULL) return 0;
    *val++='\0';
    setproperty=stringtoprop(var);
    switch(setproperty){
    case op_ipaddr:
	if (inet_pton(AF_INET, val, &convaddr)<=0) return 0;
	omapi_data_string_new(&value, 4, MDL);
	memcpy(value->value, &convaddr.s_addr,4);
	break;
    case op_hwaddr:
	if (strlen(val) != 17) return 0;
	for (i=0; i < 17; i+=3){
	    if (i != 15 && val[i+2] != ':') return 0;
	    val[i+2]='\0';
	    hwaddr[i/3]=(char)strtol(&val[i], NULL, 16);
	}
	omapi_data_string_new(&value, 6, MDL);
	memcpy(value->value, hwaddr, 6);
	break;
    case op_hwtype:
	hostint=atol(val);
	netint=htonl(hostint);
	omapi_data_string_new(&value, sizeof netint, MDL);
	memcpy(value->value, &netint, sizeof netint);
	break;
    case op_state:
	setstate=stringtostate(val);
	netint=htonl((unsigned long)setstate);
	omapi_data_string_new(&value, sizeof netint, MDL);
	memcpy(value->value, &netint, sizeof netint);
	break;
    case op_ends:
    case op_tstp:
    case op_tsfp:
    case op_cltt:
	strptime(val, "%a %b %d %T %Y", &timestruct);
	timeval=mktime(&timestruct);
	netint=htonl((unsigned long)timeval);
	omapi_data_string_new(&value, 4, MDL);
	memcpy(value->value, &netint, 4);
	break;
    case op_name:
    case op_client_hostname:
    case op_statements:
    case op_group:
	i=strlen(val);
	omapi_data_string_new(&value, i, MDL);
	memcpy(value->value, val, i);
	break;
    default:
	return 0;
    } /* end of switch */

    status=dhcpctl_set_value(object, value, var);
    if (status != ISC_R_SUCCESS){
	dhcpctl_data_string_dereference(&value, MDL);
	return -1;
    }
    return 1;
}

/* Return values:
     NULL - error converting value to string
     "string" - converted string
*/
char *valuetostring(dhcpctl_data_string *p_val, omprop proptype, char *buf, int bufsz){
    dhcpctl_data_string value;
    int i,j;
    time_t thetime;
    struct in_addr convaddr;
    char *statestring;
    uint32_t hostint,netint;

    value=*p_val;
    switch(proptype){
    case op_ends:
    case op_tstp:
    case op_tsfp:
    case op_cltt:
	memcpy(&netint, value->value, value->len);
	hostint=ntohl(netint);
	ctime_r((time_t *) &hostint,buf);
	buf[strlen(buf)-1]='\0';
	break;
    case op_ipaddr:
	memcpy(&convaddr.s_addr,value->value,4);
	if((void *)inet_ntop(AF_INET,&convaddr,buf,MAXLEN + 1)==NULL){
	    return NULL;
	}
	break;
    case op_client_hostname:
    case op_name:
	if(value->len > (bufsz-1)) return "error: name too long";
	memcpy(buf,value->value,value->len);
	buf[value->len]='\0';
	break;
    case op_statements:
 	if(value->len > (bufsz-1)) return "error: statement too long";
	memcpy(buf,value->value,value->len);
	buf[value->len]='\0';
    case op_dhcp_client_id:
    case op_hwaddr:
	if((value->len)*3 > (bufsz)) return "error: value too long";
	for(i=0,j=0;i<(value->len)-1;i++,j+=3){
	    sprintf(&buf[j],"%2.2x:",value->value[i]);
	}
	sprintf(&buf[j],"%2.2x",value->value[i]);
	break;
    case op_hwtype:
	memcpy(&netint,value->value,4);
	hostint=ntohl(netint);
	sprintf(buf,"%d",hostint);
	break;
    case op_state:
	memcpy(&netint,value->value,4);
	hostint=ntohl(netint);
	statestring=statetostring((omstate)hostint);
	strncpy(buf, statestring, bufsz);
	break;
    default:
	return NULL;
    }
    return buf;
}
